flair/models/lemmatization_model.py

```diff
@@ -0,0 +1,624 @@
+            if self.contextualized_embedding:
+                batch_size = len(effective_data_lenght)
+                all_seqs = dict()
+                for i in range(batch_size):
+                    all_seqs[i] = ""
+            else:
+                all_seqs = None
+
+            # If you use teacher forcing, the next input uses the correct value, otherwise use the predicted value as input for the next moment.
+            # The probability of using Teacher Forcing can be modified by adjusting the value of teacher_forcing_ratio.
+            use_teacher_forcing = True if random.random() < self.teacher_forcing_ratio else False
+
+            decoder_input = None
+
+            for t in range(max_lemma_lenght):
+                decoder_input = self._generate_decoder_input(decoder_input, all_seqs, batch_size)
+                decoder_output, decoder_hidden = self._decode(decoder_input, decoder_hidden, encoder_out)
+                mask_loss = self._calculate_loss(decoder_output, lemma[t], mask[t])
+                loss += mask_loss
+                # Use the correct value as input for the next moment
+                if use_teacher_forcing:
+                    decoder_input = lemma[t].view(1, -1)
+                else:
+                    _, topi = decoder_output.topk(1)
+                    decoder_input = torch.LongTensor([[topi[i][0] for i in range(batch_size)]]).to(flair.device)
+
+                # Using contextualized word embeddings requires recording the prediction results at each moment to generate word vectors.
+                if self.contextualized_embedding:
+                    for i in range(batch_size):
+                        all_seqs[i] += self.character_dictionary.get_item_for_index(decoder_input[0][i])
+
+        return loss, len(sentences)
+
+    def _encode(self, encoder_input, effective_data_lenght):
+        """
+        :param encoder_input: the output of the word embedding layer, i.e. the word vector,
+        :param effective_data_lenght: the valid length of each column.i.e. the original length of each token.
+        """
+        hidden = None
+        packed = nn.utils.rnn.pack_padded_sequence(encoder_input, effective_data_lenght)
+        outputs, hidden = self.encoder(packed, hidden)
+        outputs, _ = nn.utils.rnn.pad_packed_sequence(outputs)
+
+        outputs = outputs[:, :, :self.hidden_size] + outputs[:, :, self.hidden_size:]
+
+        return outputs, hidden
+
+    def _attn(self, hidden, encoder_outputs, method='dot'):
+        # TODO: Add another attention score
+        if method == 'dot':
+            attn_energies = self._dot_score(hidden, encoder_outputs)
+            attn_energies = attn_energies.t()
+
+        return F.softmax(attn_energies, dim=1).unsqueeze(1)
+
+    def _dot_score(self, hidden, encoder_output):
+        return torch.sum(hidden * encoder_output, dim=2)
```

Comment:
I am a bit confused about how the attention scores are calculated. Why are you doing a sum here? Shouldn't there be a dot product that directly gives the attention scores?

Reply:
Yes, because what is required is the value of the dot product as a score. But here is * (Hadamard product). So finally calculate their sum in the second dimension. 

for example, We can set input_lengths= 10.
so the size of hidden is (1, batch  , 2048). The size of  encoder_output is (input_lengths=10, batch  , 2048) 
hidden * encoder_outputs can multiply the hidden vector (2048 numbers) with the corresponding position to the 10 vectors (2048 numbers) of encoder_outputs. And the inner product needs to add up these 2048 products. So use sum in the second dimension.   The final result is (10, batch). 

In fact, this should be equivalent to using bmm.

Reply:
Ah ok, thanks for clarifying!